File: C_lab/week11/quicksort.c
#include <stdio.h>
void quicksort (int [], int, int);   //Function Prototype
int main()
{
    int list[50];                            //Regular procedure...........
    int size, i;
    printf("Enter the number of elements: ");
    scanf("%d", &size); 
    printf("Enter the elements to be sorted:\n");
    for (i = 0; i < size; i++)
    {
        scanf("%d", &list[i]);
    }                                          //upto here-------->
     quicksort(list, 0, size - 1);             //Function calling...
     printf("After applying quick sort\n");
    for (i = 0; i < size; i++)
    {
     printf("%d ", list[i]);
     }
}

void quicksort(int list[], int low, int high)    //Called function...
{
    int pivot, i, j, temp;
    if (low < high)
      {
         pivot = low;                      //Every First value in list as a pivot
         i = low;
         j = high;
         while (i < j) 
          {
             while (list[i] <= list[pivot] && i <= high)  //from left to right move when ever MAX value occurs stop while loop
              {
                  i++;
              }
             while (list[j] > list[pivot] && j >= low)    //from right to left move when ever MIN value occurs stop while loop
              {
                  j--;
               }

            if (i < j)                   //If True   swap i,j th positions

            {

                temp = list[i];

                list[i] = list[j];

                list[j] = temp;

            }

        }                   

        temp = list[j];                  // If i index crosses the j th index swap jth index value and pivot value

        list[j] = list[pivot];

        list[pivot] = temp;              // Pivot value inserted in to its exaclty value right now

        quicksort(list, low, j - 1);     //Again 0 to j-1 value calling

        quicksort(list, j + 1, high);    //  j+1 to high value calling

    }

}

File: DSA540/QuickSort.java
package DSA540;

public class QuickSort {

	public static void main(String[] args) {
		int a[]= {5,4,3,2,1};
		quick(a,0,a.length-1);
		for(int i=0;i<a.length;i++)
		{
			System.out.print(a[i]+" ");
		}
	}
	public static void quick(int a[],int low,int high)
	{
		if(low<high)
		{
			int p=partition(a,low,high);
			quick(a,low,p-1);
			quick(a,p+1,high);
		}
	}
	public static int partition(int a[],int low,int high)
	{
		int pivot=a[high];
		int i=(low-1);
		for(int j=low;j<=high-1;j++)
		{
			if(a[j]<pivot)
			{
				i++;
				a[i]=a[i]+a[j]-(a[j]=a[i]);
			}
		}
		a[i+1]=a[i+1]+a[high]-(a[high]=a[i+1]);
		return (i+1);
	}

}
